core/trino-main/src/main/java/io/trino/sql/planner/EffectivePredicateExtractor.java

```diff
@@ -389,6 +390,13 @@ public Expression visitValues(ValuesNode node, Void context)
                                 hasNull[i] = true;
                             }
                             else {
+                                if (item instanceof Block) {
```

Comment:
It would be more robust to check for the type directly:

```
(types.get(node.getOutputSymbols().get(i)) instanceof RowType
```

Also, if we're going to check for nulls, the check should be recursive. Some of the fields of a ROW may be non-null rows that contain nulls.

Reply:
Thanks @martint for the review. You are absolutely right, case like this will fail:
```sql
trino> SELECT *
    -> FROM (VALUES ROW(CAST(ROW(1, Row(1, null)) AS ROW(a integer, Row(b integer, c integer))))) t(r)
    -> JOIN (VALUES 42) u(c) ON c = r.a;
Query failed: ROW comparison not supported for fields with null elements
```

I modified the condition to skip predicate if the there's a null value or a nested Row exists.
I know this is not a perfect solution, but I had hard time to make recursive call working (since it will require re-build a ValueNode?) I'd really appreciate if you could show me if there's any easy way to do that.